File: cpp/tests/test_hamiltonian_regularizer.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

#include <gtest/gtest.h>

#include <qdk/chemistry/algorithms/hamiltonian.hpp>
#include <qdk/chemistry/algorithms/hamiltonian_regularizer.hpp>
#include <qdk/chemistry/algorithms/mc.hpp>
#include <qdk/chemistry/algorithms/scf.hpp>
#include <qdk/chemistry/utils/hamiltonian_one_norm.hpp>

#include "ut_common.hpp"

using namespace qdk::chemistry::algorithms;

class HamiltonianRegularizerTest : public ::testing::Test {};

TEST_F(HamiltonianRegularizerTest, FactoryHygiene) {
  auto available = HamiltonianRegularizerFactory::available();
  EXPECT_TRUE(std::find(available.begin(), available.end(),
                        "fermionic_low_rank") != available.end());

  auto regularizer = HamiltonianRegularizerFactory::create();
  EXPECT_EQ(regularizer->name(), "fermionic_low_rank");
  EXPECT_EQ(regularizer->type_name(), "hamiltonian_regularizer");

  auto regularizer_named =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  EXPECT_EQ(regularizer_named->name(), "fermionic_low_rank");

  EXPECT_THROW(HamiltonianRegularizerFactory::create("nonexistent"),
               std::runtime_error);
}

TEST_F(HamiltonianRegularizerTest, DefaultTruncationThresholdIsZero) {
  auto regularizer =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  EXPECT_DOUBLE_EQ(
      regularizer->settings().get<double>("df_truncation_threshold"), 0.0);
}

TEST_F(HamiltonianRegularizerTest, ThrowsOnUnrestrictedHamiltonian) {
  auto h_atom = testing::create_hydrogen_structure();
  auto scf_solver = ScfSolverFactory::create();
  scf_solver->settings().set("scf_type", std::string("auto"));
  scf_solver->settings().set("enable_gdm", true);
  auto [E_HF, wfn_HF] = scf_solver->run(h_atom, 0, 2, "cc-pvdz");

  auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
  auto ham = hamiltonian_constructor->run(wfn_HF->get_orbitals());
  ASSERT_FALSE(ham->is_restricted());

  auto regularizer =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  EXPECT_THROW(regularizer->run(ham, 1, 0), std::invalid_argument);
}

/**
 * @brief The strongest correctness check: BLISS shifts should not change the
 * physical energy of the target electron-number sector. Run exact FCI before
 * and after applying the FLR-BLISS shift and confirm the energies agree
 * to within the standard CI energy tolerance, both at the default
 * (no-truncation) setting and with an explicit truncation threshold.
 */
TEST_F(HamiltonianRegularizerTest, Water_STO3G_EnergyInvariantUnderShift) {
  auto water = testing::create_water_structure();
  auto scf_solver = ScfSolverFactory::create();
  auto [E_HF, wfn_HF] = scf_solver->run(water, 0, 1, "sto-3g");

  auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
  auto ham = hamiltonian_constructor->run(wfn_HF->get_orbitals());

  auto mc = MultiConfigurationCalculatorFactory::create();
  auto [E_before, wfn_before] = mc->run(ham, 5, 5);

  for (const double threshold : {0.0, 1e-6}) {
    auto regularizer =
        HamiltonianRegularizerFactory::create("fermionic_low_rank");
    regularizer->settings().set("df_truncation_threshold", threshold);
    auto shifted_ham = regularizer->run(ham, 5, 5);
    ASSERT_NE(shifted_ham, nullptr);

    auto mc_after = MultiConfigurationCalculatorFactory::create();
    auto [E_after, wfn_after] = mc_after->run(shifted_ham, 5, 5);

    EXPECT_NEAR(E_before, E_after, testing::ci_energy_tolerance)
        << "Energy not invariant at df_truncation_threshold=" << threshold;
  }
}

/**
 * @brief The FLR-BLISS shift should reduce (or at least not increase) the
 * fermionic double-factorization 1-norm relative to the unshifted
 * Hamiltonian.
 */
TEST_F(HamiltonianRegularizerTest, Water_STO3G_ReducesOneNorm) {
  auto water = testing::create_water_structure();
  auto scf_solver = ScfSolverFactory::create();
  auto [E_HF, wfn_HF] = scf_solver->run(water, 0, 1, "sto-3g");

  auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
  auto ham = hamiltonian_constructor->run(wfn_HF->get_orbitals());

  auto regularizer =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  auto shifted_ham = regularizer->run(ham, 5, 5);

  auto norm_after =
      qdk::chemistry::utils::hamiltonian_one_norm(*shifted_ham, 0.0);

  EXPECT_NEAR(norm_after.total, 27.590504297492, 1e-6);
}

/**
 * @brief compute_shift() + rebuild_bliss_shifted_hamiltonian() must reproduce
 * run() exactly. This locks the refactor that split the regularizer into a
 * public shift-computation step and a public, shift-agnostic rebuild step.
 */
TEST_F(HamiltonianRegularizerTest, ComputeShiftThenRebuildMatchesRun) {
  auto water = testing::create_water_structure();
  auto scf_solver = ScfSolverFactory::create();
  auto [E_HF, wfn_HF] = scf_solver->run(water, 0, 1, "sto-3g");

  auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
  auto ham = hamiltonian_constructor->run(wfn_HF->get_orbitals());

  auto regularizer =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  auto shifted_run = regularizer->run(ham, 5, 5);
  ASSERT_NE(shifted_run, nullptr);

  auto regularizer2 =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  auto shift = regularizer2->compute_shift(*ham, 5, 5);
  auto shifted_manual = rebuild_bliss_shifted_hamiltonian(*ham, shift, 10.0);
  ASSERT_NE(shifted_manual, nullptr);

  auto [h_run, h_run_beta] = shifted_run->get_one_body_integrals();
  auto [h_man, h_man_beta] = shifted_manual->get_one_body_integrals();
  (void)h_run_beta;
  (void)h_man_beta;
  EXPECT_TRUE(h_run.isApprox(h_man, 1e-12));

  auto [g_run, g_run_ab, g_run_bb] = shifted_run->get_two_body_integrals();
  auto [g_man, g_man_ab, g_man_bb] = shifted_manual->get_two_body_integrals();
  (void)g_run_ab;
  (void)g_run_bb;
  (void)g_man_ab;
  (void)g_man_bb;
  EXPECT_TRUE(g_run.isApprox(g_man, 1e-12));

  EXPECT_NEAR(shifted_run->get_core_energy(), shifted_manual->get_core_energy(),
              1e-12);
}

/**
 * @brief An unknown shift_method must be rejected by compute_shift()/run().
 */
TEST_F(HamiltonianRegularizerTest, ThrowsOnUnknownShiftMethod) {
  auto water = testing::create_water_structure();
  auto scf_solver = ScfSolverFactory::create();
  auto [E_HF, wfn_HF] = scf_solver->run(water, 0, 1, "sto-3g");

  auto hamiltonian_constructor = HamiltonianConstructorFactory::create();
  auto ham = hamiltonian_constructor->run(wfn_HF->get_orbitals());

  auto regularizer =
      HamiltonianRegularizerFactory::create("fermionic_low_rank");
  // The "shift_method" setting is constrained to the known methods, so an
  // unknown value is rejected at set-time rather than deferred to run().
  EXPECT_THROW(
      regularizer->settings().set("shift_method", std::string("nonexistent")),
      std::invalid_argument);
}

File: python/tests/test_hamiltonian_regularizer.py
"""Tests for the HamiltonianRegularizer (FLR-BLISS) algorithm."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import numpy as np
import pytest

from qdk_chemistry import algorithms
from qdk_chemistry.constants import ANGSTROM_TO_BOHR
from qdk_chemistry.data import Structure
from qdk_chemistry.utils import double_factorize, hamiltonian_one_norm

from .reference_tolerances import (
    ci_energy_tolerance,
    float_comparison_relative_tolerance,
)


def create_water_structure():
    """Create a water molecule structure.

    Crawford geometry - same as used in C++ tests.
    """
    symbols = ["O", "H", "H"]
    coords = (
        np.array(
            [
                [0.000000000, -0.0757918436, 0.000000000000],
                [0.866811829, 0.6014357793, -0.000000000000],
                [-0.866811829, 0.6014357793, -0.000000000000],
            ]
        )
        * ANGSTROM_TO_BOHR
    )
    return Structure(symbols, coords)


@pytest.fixture(scope="module")
def water_hamiltonian():
    """Build a water/STO-3G Hamiltonian, shared across tests in this module."""
    water = create_water_structure()
    scf_solver = algorithms.create("scf_solver")
    _, wfn_hf = scf_solver.run(water, 0, 1, "sto-3g")
    ham_constructor = algorithms.create("hamiltonian_constructor")
    return ham_constructor.run(wfn_hf.get_orbitals())


class TestHamiltonianRegularizerFactory:
    """Test factory registration and settings hygiene."""

    def test_factory(self):
        available = algorithms.available("hamiltonian_regularizer")
        assert isinstance(available, list)
        assert "fermionic_low_rank" in available

        regularizer = algorithms.create("hamiltonian_regularizer")
        assert regularizer is not None
        assert regularizer.name() == "fermionic_low_rank"

        regularizer_named = algorithms.create("hamiltonian_regularizer", "fermionic_low_rank")
        assert regularizer_named.name() == "fermionic_low_rank"

        with pytest.raises(KeyError):
            algorithms.create("hamiltonian_regularizer", "nonexistent")

    def test_default_truncation_threshold_is_zero(self):
        regularizer = algorithms.create("hamiltonian_regularizer", "fermionic_low_rank")
        assert regularizer.settings().get("df_truncation_threshold") == 0.0


class TestHamiltonianRegularizerCorrectness:
    """Physics correctness tests: energy invariance and 1-norm reduction."""

    def test_energy_invariant_under_shift(self, water_hamiltonian):
        """The correctness check: FCI energy cannot change after FLR-BLISS shift."""

        mc = algorithms.create("multi_configuration_calculator", "macis_cas")
        e_before, _ = mc.run(water_hamiltonian, 5, 5)

        for threshold in (0.0, 1e-6):
            regularizer = algorithms.create("hamiltonian_regularizer", "fermionic_low_rank")
            regularizer.settings().set("df_truncation_threshold", threshold)
            shifted_ham = regularizer.run(water_hamiltonian, 5, 5)
            assert shifted_ham is not None

            mc_after = algorithms.create("multi_configuration_calculator", "macis_cas")
            e_after, _ = mc_after.run(shifted_ham, 5, 5)

            assert np.isclose(
                e_before,
                e_after,
                rtol=float_comparison_relative_tolerance,
                atol=ci_energy_tolerance,
            ), f"Energy not invariant at df_truncation_threshold={threshold}"

    def test_reduces_one_norm(self, water_hamiltonian):
        norm_before = hamiltonian_one_norm(water_hamiltonian, 0.0)

        regularizer = algorithms.create("hamiltonian_regularizer", "fermionic_low_rank")
        shifted_ham = regularizer.run(water_hamiltonian, 5, 5)

        norm_after = hamiltonian_one_norm(shifted_ham, 0.0)

        assert norm_after.total <= norm_before.total + 1e-10


class TestDoubleFactorizationUtils:
    """Tests double_factorize and hamiltonian_one_norm."""

    def test_double_factorize_default_no_truncation(self, water_hamiltonian):
        g_aaaa, _, _ = water_hamiltonian.get_two_body_integrals()
        norb = water_hamiltonian.get_orbitals().get_num_molecular_orbitals()

        fragments_default = double_factorize(g_aaaa, norb)
        fragments_explicit_zero = double_factorize(g_aaaa, norb, 0.0)
        assert len(fragments_default) == len(fragments_explicit_zero)

        fragments_loose = double_factorize(g_aaaa, norb, 1e-2)
        assert len(fragments_loose) <= len(fragments_explicit_zero)

    def test_hamiltonian_one_norm_standalone(self, water_hamiltonian):
        norm = hamiltonian_one_norm(water_hamiltonian)
        assert norm.one_body > 0.0
        assert norm.two_body > 0.0
        assert np.isclose(norm.total, norm.one_body + norm.two_body)
